addons/addon-base-raas/packages/base-raas-services/lib/environment/service-catalog/environment-sc-cidr-service.js

```diff
@@ -0,0 +1,221 @@
+        securityGroupId,
+        currentIngressRules,
+        updateRequest,
+      });
+
+      // Not storing the changes in the DB, but since all went well
+      // we return the cidr field as part of the env obj
+      existingEnvironment.cidr = newCidrList;
+    });
+
+    // Write audit event
+    await this.audit(requestContext, { action: 'update-environment-sc-cidr', body: existingEnvironment });
+
+    return existingEnvironment;
+  }
+
+  getIpPermission(updateRule, cidrList) {
+    const ipv4List = _.filter(cidrList, cidr => IsCidr.v4(cidr));
+    // const ipv6List = _.filter(cidrList, cidr => IsCidr.v6(cidr));
+
+    const IpPermission = {
+      FromPort: updateRule.fromPort,
+      IpProtocol: updateRule.protocol,
+      IpRanges: _.map(ipv4List, cidr => ({
+        CidrIp: cidr,
+        // For future: Custom description doesn't work, since that has to match exactly what's there in the SG.
+        // This would require storing cidr field in DB
+        // Description: `Updated via SWB`,
+      })),
+      // TODO: Ipv6 calls have to be done separately and cannot habe
+      // Ipv6Ranges: _.map(ipv6List, cidr => ({
+      //   CidrIp: cidr,
+      //   Description: `Updated via SWB`,
+      // })),
+      ToPort: updateRule.toPort,
+    };
+    return IpPermission;
+  }
+
+  // This method is smart enough to determine which IPs to revoke/grant access
+  // by peeking into the ingress rules of the workspace's security group
+  async manageIngressRules(
+    requestContext,
+    { existingEnvironment, securityGroupId, currentIngressRules, updateRequest } = {},
+  ) {
+    const addIpPermissions = [];
+    const revokeIpPermissions = [];
+    const newCidrList = [];
+
+    _.forEach(currentIngressRules, existingRule => {
+      const matchingRule = _.find(updateRequest, {
+        fromPort: existingRule.fromPort,
+        toPort: existingRule.toPort,
+        protocol: existingRule.protocol,
+      });
+      if (matchingRule) {
+        const addCidrs = _.difference(matchingRule.cidrBlocks, existingRule.cidrBlocks);
+        const revokeCidrs = _.difference(existingRule.cidrBlocks, matchingRule.cidrBlocks);
+
+        if (!_.isEmpty(addCidrs)) addIpPermissions.push(this.getIpPermission(matchingRule, addCidrs));
+        if (!_.isEmpty(revokeCidrs)) revokeIpPermissions.push(this.getIpPermission(matchingRule, revokeCidrs));
+        newCidrList.push(matchingRule);
+      }
+    });
+
+    // Get params ready
+    const revokeParams = { GroupId: securityGroupId, IpPermissions: revokeIpPermissions };
+    const authorizeParams = { GroupId: securityGroupId, IpPermissions: addIpPermissions };
+
+    // Start making security group updates
+    const ec2Client = await this.getEc2Client(requestContext, existingEnvironment.id);
+    if (!_.isEmpty(revokeIpPermissions)) await this.revokeSecurityGroupIngress(ec2Client, revokeParams);
+    if (!_.isEmpty(addIpPermissions)) await this.authorizeSecurityGroupIngress(ec2Client, authorizeParams);
```

Comment:
Since there are multiple networking calls involved, it is possible that we have a partial update scenario and throw an exception half way through the update. How does UI handle that. Will it display the partial update correctly? Would calling update once simplify things for us?

Reply:
There are at most only two calls made for any given update request, revoke and authorize. If there is any error during these calls, the UI throws an error and closes the Edit CIDRs component. If the user wishes to try again, they have to click on this component which fetches the latest state (partially updated) of the ingress rules.

Reply:
ok, so the user will see the latest state during partial updates? Have you tested the experience by deliberately making say the authorize call fail?

Reply:
Yes. In case of partial updates, the user sees an error dialogue. The last updated Ingress Rule state is displayed when the UI Store (holding the value of the CIDR blocks) is updated (once every 30 sec), or when the page refreshes, whichever happens first.